exporter/exporterhelper/common.go

```diff
@@ -27,195 +26,213 @@ type baseRequestSender struct {
 	consumerOptions []consumer.Option
 }
 
 // TODO: requestExporter, marshaler, and unmarshaler arguments can be removed when the old exporter helpers will be updated to call the new ones.
 func newBaseExporter(set exporter.CreateSettings, signal component.DataType, requestExporter bool, marshaler RequestMarshaler,
-	unmarshaler RequestUnmarshaler, osf obsrepSenderFactory, options ...Option) (*baseExporter, error) {
+	unmarshaler RequestUnmarshaler, osf obsrepSenderFactory, opts ...Option) (*baseExporter, error) {
 
 	obsReport, err := NewObsReport(ObsReportSettings{ExporterID: set.ID, ExporterCreateSettings: set})
 	if err != nil {
 		return nil, err
 	}
 
-	be := &baseExporter{
+	sets := &settings{
 		requestExporter: requestExporter,
-		marshaler:       marshaler,
-		unmarshaler:     unmarshaler,
-		signal:          signal,
-
-		queueSender:   &baseRequestSender{},
-		obsrepSender:  osf(obsReport),
-		retrySender:   &baseRequestSender{},
-		timeoutSender: &timeoutSender{cfg: NewDefaultTimeoutSettings()},
-
-		set:    set,
-		obsrep: obsReport,
+		timeoutConfig:   NewDefaultTimeoutSettings(),
 	}
-
-	for _, op := range options {
-		op(be)
+	for _, op := range opts {
+		op(sets)
 	}
-	be.connectSenders()
 
-	// If retry sender is disabled then disable requeuing in the queue sender.
-	// TODO: Make re-enqueuing configurable on queue sender instead of relying on retry sender.
-	if qs, ok := be.queueSender.(*queueSender); ok {
-		// if it's not retrySender, then it is disabled.
-		if _, ok = be.retrySender.(*retrySender); !ok {
-			qs.requeuingEnabled = false
+	be := &baseExporter{
+		obsrep:          obsReport,
+		consumerOptions: sets.consumerOptions,
+	}
+	// Create senders in reverse order:
+	// Start with the basic exporter wrapper.
+	be.exporterSender = &exporterSender{StartFunc: sets.startFunc, ShutdownFunc: sets.shutdownFunc}
+	be.addSender(be.exporterSender)
+	// Before that add the timeout sender.
+	be.addSender(newTimeoutSender(sets.timeoutConfig, be.lastSender))
+	// Before that add the retry sender if configured.
+	if sets.retryConfig != nil {
+		if !sets.retryConfig.Enabled {
+			// TODO: Simplify logic for error logging request. Consider to add it always if both retry_on_failure is not enabled, not only when configured but disabled.
+			be.addSender(&errorLoggingRequestSender{
+				baseRequestSender: baseRequestSender{nextSender: be.lastSender},
+				logger:            set.Logger,
+				message:           "Exporting failed. Try enabling retry_on_failure config option to retry on retryable errors",
+			})
+		} else {
+			be.retrySender = newRetrySender(*sets.retryConfig, set, be.lastSender)
+			be.addSender(be.retrySender)
+		}
+	}
+	// Before that add the observability sender.
+	be.addSender(osf(obsReport, be.lastSender))
+	// Before that add the queue sender if configured.
+	if sets.queueConfig != nil {
+		if !sets.queueConfig.Enabled {
+			// TODO: Consider to remove this. Not sure how this helps with dropping data.
```

Comment:
I think we can remove this once https://github.com/open-telemetry/opentelemetry-collector/pull/9090 is merged